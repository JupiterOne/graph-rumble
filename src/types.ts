// Providers often supply types with their API libraries.

import {
  IntegrationInstance,
  IntegrationLogger,
} from '@jupiterone/integration-sdk-core';
import { IntegrationConfig } from './config';

export interface RunZeroAccount {
  id: string;
  name: string;
}

export interface RunZeroOrganization {
  id: string;
  name: string;
  created_at: number;
  created_by: string;
  updated_at: number;
  client_id: string;
  download_token: string;
  download_token_created_at: number;
  demo: boolean;
  project: boolean;
  parent_id: string;
  description: string | null;
  inactive: boolean;
  deactivated_at: number;
  service_count: number;
  service_count_tcp: number;
  service_count_udp: number;
  service_count_arp: number;
  service_count_icmp: number;
  asset_count: number;
  live_asset_count: number;
  recent_asset_count: number;
  software_count: number;
  vulnerability_count: number;
  export_token: string;
  export_token_created_at: number;
  export_token_last_used_at: number;
  export_token_last_used_by: string;
  export_token_counter: number;
  expiration_assets_stale: number;
  expiration_assets_offline: number;
  expiration_scans: number;
  expiration_warning_last_sent: number;
}

export interface RunZeroUser {
  id: string;
  client_id: string;
  created_at: number;
  updated_at: number;
  first_name: string | null;
  last_name: string | null;
  email: string | null;
  client_admin: boolean;
  // the first user can have a role of '' in this case that user will also be a client_admin
  // and will be assigned a role of 'admin' on their USER_ASSIGNED_ORGANIZATION
  org_default_role: Role | '' | null;
  org_roles: Record<string, Role>;
  reset_token_expiration: number;
  invite_token_expiration: number;
  last_login_ip: string | null;
  last_login_at: number;
  last_login_ua: string | null;
  last_activity_at: number;
  sso_only: boolean;
  login_failures: number;
  actions: number;
  last_action_at: number;
}

export interface RunZeroSite {
  id: string;
  created_at: number;
  updated_at: number;
  client_id: string;
  organization_id: string;
  name: string | null;
  description: string | null;
  scope: string | null;
  excludes: string | null;
  inactive: boolean;
  deactivated_at: number;
  service_count: number;
  service_count_tcp: number;
  service_count_udp: number;
  service_count_arp: number;
  service_count_icmp: number;
  asset_count: number;
  live_asset_count: number;
  recent_asset_count: number;
  software_count: number;
  vulnerability_count: number;
  subnets: Record<string, string | null>;
  asset_address_count: number;
  asset_address_extra_count: number;
  last_task_id: string | null;
  last_task_at: number;
  last_task_by: string | null;
  last_task_duration: number;
}

export interface RunZeroAsset {
  id: string;
  created_at: number;
  updated_at: number;
  organization_id: string;
  site_id: string;
  alive: boolean;
  first_seen: number;
  last_seen: number;
  detected_by?: string;
  type?: string;
  os?: string;
  os_vendor?: string;
  os_product?: string;
  os_version?: string;
  hw?: string;
  hw_vendor?: string;
  hw_product?: string;
  hw_version?: string;
  addresses?: string[];
  addresses_extra?: string[];
  macs?: string[];
  mac_vendors?: string[];
  names?: string[];
  domains?: string[];
  tags?: Record<string, any>;
  // TODO: better typings
  // Documentation schema not updated
  services?: Record<string, Object>;
  credentials: Object;
  rtts: Record<string, string[]>;
  attributes: Record<string, string>;
  service_count: number;
  service_count_tcp: number;
  service_count_udp: number;
  service_count_arp: number;
  service_count_icmp: number;
  software_count: number;
  vulnerability_count: number;
  lowest_ttl: number;
  lowest_rtt: number;
  last_agent_id: string;
  last_task_id: string;
  newest_mac: string;
  newest_mac_vendor: string;
  newest_mac_age: number;
  service_ports_tcp: string[];
  service_ports_udp: string[];
  service_protocols: string[];
  service_products: string[];
  scanned: boolean;
  source_ids: number[];
  eol_os: number;
  eol_os_ext: number;
  outlier_score: number;
  outlier_raw: number;
  sources: string[];
  org_name: string;
  site_name: string;
  agent_name: string | null;
  agent_external_ip: string | null;
  hosted_zone_name: string | null;
  subnets: Record<string, string | null>;
  foreign_attributes: Object;
}

export type APIClientOptions = {
  instance: IntegrationInstance<IntegrationConfig>;
  name: string;
  logger: IntegrationLogger;
};

export type Role = 'none' | 'viewer' | 'annotator' | 'user' | 'admin';
export const RoleLevel = {
  none: 0,
  viewer: 1,
  annotator: 2,
  user: 3,
  admin: 4,
};
