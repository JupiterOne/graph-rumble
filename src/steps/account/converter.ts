import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';
import { RunZeroAccount } from '../../types';

import { Entities } from '../constants';
export function createAccountEntity(account: RunZeroAccount): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: account.id,
        name: account.name,
      },
      assign: {
        _key: account.id,
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        name: account.name,
        displayName: account.name,
      },
    },
  });
}
