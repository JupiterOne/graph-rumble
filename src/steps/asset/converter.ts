import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { RunZeroAsset } from '../../types';
import { Entities } from '../constants';

export function createAssetEntity(asset: RunZeroAsset): Entity {
  const name = selectNameForAsset(asset);

  return createIntegrationEntity({
    entityData: {
      source: asset,
      assign: {
        _key: asset.id,
        _type: Entities.ASSET._type,
        _class: Entities.ASSET._class,
        name: name,
        createdAt: parseTimePropertyValue(asset.created_at, 'ms'),
        updatedAt: parseTimePropertyValue(asset.updated_at, 'ms'),
        alive: asset.alive,
        lastSeen: parseTimePropertyValue(asset.last_seen, 'ms'),
        firstSeen: parseTimePropertyValue(asset.first_seen, 'ms'),
        detectedBy: asset.detected_by,
        type: asset.type,
        osVendor: asset.os_vendor,
        hardware: asset.hw,
        category: asset.type,
        make: asset.hw_vendor,
        model: asset.hw_product,
        version: asset.hw_version,
        serial: null,
        hardwareVendor: asset.hw_vendor,
        hardwareModel: asset.hw_product,
        hardwareVersion: asset.hw_version,
        osDetails: asset.os + (asset.os_version ? ' ' + asset.os_version : ''),
        osName: asset.os_product,
        osVersion: asset.os_version,
        macAddresses: asset.macs,
        deviceId: null,
        serviceCount: asset.service_count,
        softwareCount: asset.software_count,
        vulnerabilityCount: asset.vulnerability_count,
        lastAgentId: asset.last_agent_id,
        lastTaskId: asset.last_task_id,
        newestMacAddress: asset.newest_mac,
        newestMacVendor: asset.newest_mac_vendor,
        orgName: asset.org_name,
        siteName: asset.site_name,
      },
    },
  });
}

function selectNameForAsset(asset: RunZeroAsset) {
  if (asset.names && asset.names.length > 0) {
    return asset.names[0];
  } else if (asset.hw) {
    return asset.hw;
  } else {
    return asset.id;
  }
}
