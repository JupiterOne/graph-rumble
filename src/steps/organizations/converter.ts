import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { RunZeroOrganization } from '../../types';
import { Entities } from '../constants';

export function createOrganizationEntity(org: RunZeroOrganization): Entity {
  return createIntegrationEntity({
    entityData: {
      // for source we could pull in all the non-secret fields
      source: {
        ...org,
        download_token: undefined,
        export_token: undefined,
      },
      assign: {
        _key: org.id,
        _type: Entities.ORGANIZATION._type,
        _class: Entities.ORGANIZATION._class,
        name: org.name,
        displayName: org.name ?? undefined,
        // RunZeroOrganization specific properties
        clientId: org.client_id,
        createdAt: parseTimePropertyValue(org.created_at, 'ms'),
        createdBy: org.created_by,
        updatedAt: parseTimePropertyValue(org.updated_at, 'ms'),
        downloadTokenCreatedAt: parseTimePropertyValue(
          org.download_token_created_at,
          'ms',
        ),
        demo: org.demo,
        project: org.project,
        parentId: org.parent_id,
        // description can be a null field, if it is coerce to empty string
        description: org.description ?? undefined,
        inactive: org.inactive,
        deactivatedAt: parseTimePropertyValue(org.deactivated_at, 'ms'),
        serviceCount: org.service_count,
        serviceCountTCP: org.service_count_tcp,
        serviceCountUDP: org.service_count_udp,
        serviceCountARP: org.service_count_arp,
        serviceCountICMP: org.service_count_icmp,
        assetCount: org.asset_count,
        liveAssetCount: org.live_asset_count,
        recentAssetCount: org.recent_asset_count,
        softwareCount: org.software_count,
        vulnerabilityCount: org.vulnerability_count,
        exportTokenCreatedAt: parseTimePropertyValue(
          org.export_token_created_at,
          'ms',
        ),
        exportTokenLastUsedAt: parseTimePropertyValue(
          org.export_token_last_used_at,
          'ms',
        ),
        exportTokenLastUsedBy: org.export_token_last_used_by,
        exportTokenCounter: org.export_token_counter,
        expirationAssetsStale: org.expiration_assets_stale,
        expirationAssetsOffline: org.expiration_assets_offline,
        expirationScans: org.expiration_scans,
        expirationWarningLastSent: parseTimePropertyValue(
          org.expiration_warning_last_sent,
          'ms',
        ),
        // Fields that are on RunZeroOrganization but excluded from assign:
        // download_token
        // export_token
      },
    },
  });
}
