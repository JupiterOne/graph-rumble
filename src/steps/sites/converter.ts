import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { RunZeroSite } from '../../types';
import { Entities } from '../constants';

export function createSiteEntity(site: RunZeroSite): Entity {
  return createIntegrationEntity({
    entityData: {
      source: site,
      assign: {
        _class: Entities.SITE._class,
        _key: site.id,
        _type: Entities.SITE._type,
        createdAt: parseTimePropertyValue(site.created_at, 'ms'),
        updatedAt: parseTimePropertyValue(site.updated_at, 'ms'),
        clientId: site.client_id,
        organizationId: site.organization_id,
        name: site.name ?? undefined,
        description: site.description ?? undefined,
        scope: site.scope ?? undefined,
        excludes: site.excludes ?? undefined,
        inactive: site.inactive,
        deactivatedAt: parseTimePropertyValue(site.deactivated_at, 'ms'),
        serviceCount: site.service_count,
        serviceCountTCP: site.service_count_tcp,
        serviceCountUDP: site.service_count_udp,
        serviceCountARP: site.service_count_arp,
        serviceCountICMP: site.service_count_icmp,
        assetCount: site.asset_count,
        liveAssetCount: site.live_asset_count,
        recentAssetCount: site.recent_asset_count,
        softwareCount: site.software_count,
        vulnerabilityCount: site.vulnerability_count,
        assetAddressCount: site.asset_address_count,
        assetAddressExtraCount: site.asset_address_extra_count,
        lastTaskId: site.last_task_id ?? undefined,
        lastTaskAt: parseTimePropertyValue(site.last_task_at, 'ms'),
        lastTaskBy: site.last_task_by ?? undefined,
        lastTaskDuration: site.last_task_duration,
      },
    },
  });
}
