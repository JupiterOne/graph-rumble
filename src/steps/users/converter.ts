import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { RunZeroUser } from '../../types';
import { Entities } from '../constants';

export function createUserEntity(user: RunZeroUser): Entity {
  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: createAssignObject(user),
    },
  });
}

function createAssignObject(user: RunZeroUser) {
  let name: string | undefined = undefined;
  if (user.first_name && user.last_name) {
    name = user.first_name + ' ' + user.last_name;
  } else if (user.first_name) {
    name = user.first_name;
  } else if (user.last_name) {
    name = user.last_name;
  }

  return {
    _key: user.id,
    _type: Entities.USER._type,
    _class: Entities.USER._class,
    name: name,
    displayName: name,
    // we add the admin derived admin property to make querying easier
    // and to resemble other graphs
    admin: user.client_admin === true || user.org_default_role === 'admin',
    // okay for properties to be undefined, but not okay to be null
    firstName: user.first_name ?? undefined,
    lastName: user.last_name ?? undefined,
    username: user.email ?? undefined,
    email: user.email ?? undefined,
    orgDefaultRole: user.org_default_role ?? undefined,
    resetTokenExpiration: parseTimePropertyValue(
      user.reset_token_expiration,
      'ms',
    ),
    inviteTokenExpiration: parseTimePropertyValue(
      user.invite_token_expiration,
      'ms',
    ),
    lastLoginIP: user.last_login_ip ?? undefined,
    lastLoginAt: parseTimePropertyValue(user.last_login_at, 'ms'),
    lastLoginUa: user.last_login_ua ?? undefined,
    lastActivityAt: parseTimePropertyValue(user.last_activity_at, 'ms'),
    ssoOnly: user.sso_only,
    loginFailures: user.login_failures,
    actions: user.actions,
    lastActionAt: parseTimePropertyValue(user.last_action_at, 'ms'),
    active: true,
  };
}
