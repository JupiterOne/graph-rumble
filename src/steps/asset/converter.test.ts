import { RunZeroAsset } from '../../types';
import { createAssetEntity } from './converter';

describe('#createAssetEntity', () => {
  test('should convert to entity', () => {
    const asset: RunZeroAsset = {
      id: '40fa9e9a-e798-44e4-a521-d06a7782ed14',
      created_at: 1642795487,
      updated_at: 1642795487,
      organization_id: '463fed64-54bf-4369-b4b4-86a039773470',
      site_id: 'c9f0bc41-afdb-4053-945a-2a5333d1565a',
      alive: true,
      last_seen: 1583271232,
      first_seen: 1583271138,
      detected_by: 'arp',
      type: 'Server',
      os_vendor: 'Ubuntu',
      os_product: 'Linux',
      os_version: '',
      os: 'Ubuntu Linux',
      hw_vendor: '',
      hw_product: '',
      hw_version: '',
      hw: '',
      addresses: ['192.168.0.1'],
      addresses_extra: [],
      macs: ['a0:36:9f:50:77:d8'],
      mac_vendors: ['Intel Corporate'],
      names: ['ROUTER'],
      domains: [],
      services: {
        '192.168.0.1/0/arp/': {
          mac: 'a0:36:9f:50:77:d8',
          macDateAdded: '2011-04-29',
          macVendor: 'Intel Corporate',
          'service.address': '192.168.0.1',
          'service.id': '40fa9e9a-e798-44e4-a521-d06a7782ed14',
          'service.port': '0',
          'service.transport': 'arp',
          source:
            'arp\tsnmp.arpcache.192.168.0.19\tsnmp.arpcache.192.168.0.30\tsnmp.arpcache.192.168.0.31\tsnmp.arpcache.192.168.0.32',
          ts: '1583271232',
        },
        '192.168.0.1/0/icmp/': {
          addrs: '192.168.0.1',
          rtts: '104457\t78781',
          'service.address': '192.168.0.1',
          'service.id': '40fa9e9a-e798-44e4-a521-d06a7782ed14',
          'service.port': '0',
          'service.transport': 'icmp',
          ts: '1583271150',
        },
        '192.168.0.1/53/udp/': {
          addrs: '192.168.0.1',
          'fp.certainty': '0.85',
          'hostname.bind': 'router',
          'os.cpe23': 'cpe:/o:canonical:ubuntu_linux:-',
          'os.family': 'Linux',
          'os.product': 'Linux',
          'os.vendor': 'Ubuntu',
          product: 'ISC:BIND:9.11.3',
          protocol: 'dns',
          replies: 'www.google.com=A,172.217.1.132',
          resolvers: '172.253.212.13',
          rtts: '158454077\t93812355',
          'service.address': '192.168.0.1',
          'service.cpe23': 'cpe:/a:isc:bind:9.11.3',
          'service.family': 'BIND',
          'service.id': '40fa9e9a-e798-44e4-a521-d06a7782ed14',
          'service.port': '53',
          'service.transport': 'udp',
          subnets: '1.2.3.0/24-f=1-c=0-s=0',
          ts: '1583271194',
          'version.bind': '9.11.3-1ubuntu1.11-Ubuntu',
        },
        '192.168.0.1/5353/udp/': {
          protocol: 'mdns',
          replies: '1.0.168.192.in-addr.arpa.=PTR,router.local.',
          'service.address': '192.168.0.1',
          'service.id': '40fa9e9a-e798-44e4-a521-d06a7782ed14',
          'service.port': '5353',
          'service.transport': 'udp',
          ts: '1583271140',
        },
      },
      credentials: {},
      rtts: {
        'icmp/echo': ['78781', '104457'],
        'udp/dns': ['93812355', '158454077'],
      },
      attributes: {
        '@rumble.scan._links.ports.connected':
          '10:da:43:d7:59:14-192.168.0.31-1/0/12',
        '@rumble.scan._macs.ipmap': 'a0:36:9f:50:77:d8=192.168.0.1',
        '@rumble.scan.fp.certainty': '0.85',
        '@rumble.scan.match.db': 'dns-versionbind',
        '@rumble.scan.match.score': '00',
        '@rumble.scan.match.score.hw': '00',
        '@rumble.scan.match.score.os': '85',
        '@rumble.scan.os.cpe23': 'cpe:/o:canonical:ubuntu_linux:-',
        '@rumble.scan.os.family': 'Linux',
        '@rumble.scan.os.product': 'Linux',
        '@rumble.scan.os.vendor': 'Ubuntu',
        '@rumble.scan.service.cpe23': 'cpe:/a:isc:bind:9.11.3',
        '@rumble.scan.service.family': 'BIND',
        '@rumble.scan.service.product': 'BIND',
        '@rumble.scan.service.vendor': 'ISC',
        '@rumble.scan.service.version': '9.11.3',
        '@rumble.scan.switch.ip': '192.168.0.31',
        '@rumble.scan.switch.name': 'M4300-OFFICE',
        '@rumble.scan.switch.port': '192.168.0.31-1/0/12',
        '@rumble.scan.vlan': '1',
        '_links.ports.connected': '10:da:43:d7:59:14-192.168.0.31-1/0/12',
        '_macs.ipmap': 'a0:36:9f:50:77:d8=192.168.0.1',
        'fp.certainty': '0.85',
        'match.db': 'dns-versionbind',
        'match.score': '00',
        'match.score.hw': '00',
        'match.score.os': '85',
        'os.cpe23': 'cpe:/o:canonical:ubuntu_linux:-',
        'os.family': 'Linux',
        'os.product': 'Linux',
        'os.vendor': 'Ubuntu',
        'service.cpe23': 'cpe:/a:isc:bind:9.11.3',
        'service.family': 'BIND',
        'service.product': 'BIND',
        'service.vendor': 'ISC',
        'service.version': '9.11.3',
        'switch.ip': '192.168.0.31',
        'switch.name': 'M4300-OFFICE',
        'switch.port': '192.168.0.31-1/0/12',
        vlan: '1',
      },
      service_count: 4,
      service_count_tcp: 0,
      service_count_udp: 2,
      service_count_arp: 1,
      service_count_icmp: 1,
      lowest_ttl: 0,
      lowest_rtt: 78781,
      last_agent_id: '00000000-0000-0000-0000-000000000000',
      last_task_id: '374cc0f8-ca77-40d8-96c8-b53d563e7880',
      newest_mac: 'a0:36:9f:50:77:d8',
      newest_mac_vendor: 'Intel Corporate',
      newest_mac_age: 1304035200000000000,
      service_ports_tcp: [],
      service_ports_udp: ['53', '5353'],
      service_protocols: ['dns', 'mdns'],
      service_products: ['isc bind'],
      scanned: true,
      source_ids: [1],
      eol_os: 0,
      eol_os_exit: 0,
      sources: ['rumble'],
      org_name: 'Demo Organization',
      site_name: 'Corporate',
      agent_name: null,
      subnets: {},
      software_count: 0,
      vulnerability_count: 0,
      outlier_score: 0,
      outlier_raw: 0,
      agent_external_ip: null,
      hosted_zone_name: null,
      foreign_attributes: {},
    };
    expect(createAssetEntity(asset)).toMatchSnapshot();
  });
});
