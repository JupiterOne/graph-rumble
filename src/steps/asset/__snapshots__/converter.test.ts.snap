// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createAssetEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "Device",
  ],
  "_key": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "addresses": Array [
          "192.168.0.1",
        ],
        "addresses_extra": Array [],
        "agent_external_ip": null,
        "agent_name": null,
        "alive": true,
        "attributes": Object {
          "@rumble.scan._links.ports.connected": "10:da:43:d7:59:14-192.168.0.31-1/0/12",
          "@rumble.scan._macs.ipmap": "a0:36:9f:50:77:d8=192.168.0.1",
          "@rumble.scan.fp.certainty": "0.85",
          "@rumble.scan.match.db": "dns-versionbind",
          "@rumble.scan.match.score": "00",
          "@rumble.scan.match.score.hw": "00",
          "@rumble.scan.match.score.os": "85",
          "@rumble.scan.os.cpe23": "cpe:/o:canonical:ubuntu_linux:-",
          "@rumble.scan.os.family": "Linux",
          "@rumble.scan.os.product": "Linux",
          "@rumble.scan.os.vendor": "Ubuntu",
          "@rumble.scan.service.cpe23": "cpe:/a:isc:bind:9.11.3",
          "@rumble.scan.service.family": "BIND",
          "@rumble.scan.service.product": "BIND",
          "@rumble.scan.service.vendor": "ISC",
          "@rumble.scan.service.version": "9.11.3",
          "@rumble.scan.switch.ip": "192.168.0.31",
          "@rumble.scan.switch.name": "M4300-OFFICE",
          "@rumble.scan.switch.port": "192.168.0.31-1/0/12",
          "@rumble.scan.vlan": "1",
          "_links.ports.connected": "10:da:43:d7:59:14-192.168.0.31-1/0/12",
          "_macs.ipmap": "a0:36:9f:50:77:d8=192.168.0.1",
          "fp.certainty": "0.85",
          "match.db": "dns-versionbind",
          "match.score": "00",
          "match.score.hw": "00",
          "match.score.os": "85",
          "os.cpe23": "cpe:/o:canonical:ubuntu_linux:-",
          "os.family": "Linux",
          "os.product": "Linux",
          "os.vendor": "Ubuntu",
          "service.cpe23": "cpe:/a:isc:bind:9.11.3",
          "service.family": "BIND",
          "service.product": "BIND",
          "service.vendor": "ISC",
          "service.version": "9.11.3",
          "switch.ip": "192.168.0.31",
          "switch.name": "M4300-OFFICE",
          "switch.port": "192.168.0.31-1/0/12",
          "vlan": "1",
        },
        "created_at": 1642795487,
        "credentials": Object {},
        "detected_by": "arp",
        "domains": Array [],
        "eol_os": 0,
        "eol_os_ext": 0,
        "first_seen": 1583271138,
        "foreign_attributes": Object {},
        "hosted_zone_name": null,
        "hw": "",
        "hw_product": "",
        "hw_vendor": "",
        "hw_version": "",
        "id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
        "last_agent_id": "00000000-0000-0000-0000-000000000000",
        "last_seen": 1583271232,
        "last_task_id": "374cc0f8-ca77-40d8-96c8-b53d563e7880",
        "lowest_rtt": 78781,
        "lowest_ttl": 0,
        "mac_vendors": Array [
          "Intel Corporate",
        ],
        "macs": Array [
          "a0:36:9f:50:77:d8",
        ],
        "names": Array [
          "ROUTER",
        ],
        "newest_mac": "a0:36:9f:50:77:d8",
        "newest_mac_age": 1304035200000000000,
        "newest_mac_vendor": "Intel Corporate",
        "org_name": "Demo Organization",
        "organization_id": "463fed64-54bf-4369-b4b4-86a039773470",
        "os": "Ubuntu Linux",
        "os_product": "Linux",
        "os_vendor": "Ubuntu",
        "os_version": "",
        "outlier_raw": 0,
        "outlier_score": 0,
        "rtts": Object {
          "icmp/echo": Array [
            "78781",
            "104457",
          ],
          "udp/dns": Array [
            "93812355",
            "158454077",
          ],
        },
        "scanned": true,
        "service_count": 4,
        "service_count_arp": 1,
        "service_count_icmp": 1,
        "service_count_tcp": 0,
        "service_count_udp": 2,
        "service_ports_tcp": Array [],
        "service_ports_udp": Array [
          "53",
          "5353",
        ],
        "service_products": Array [
          "isc bind",
        ],
        "service_protocols": Array [
          "dns",
          "mdns",
        ],
        "services": Object {
          "192.168.0.1/0/arp/": Object {
            "mac": "a0:36:9f:50:77:d8",
            "macDateAdded": "2011-04-29",
            "macVendor": "Intel Corporate",
            "service.address": "192.168.0.1",
            "service.id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
            "service.port": "0",
            "service.transport": "arp",
            "source": "arp	snmp.arpcache.192.168.0.19	snmp.arpcache.192.168.0.30	snmp.arpcache.192.168.0.31	snmp.arpcache.192.168.0.32",
            "ts": "1583271232",
          },
          "192.168.0.1/0/icmp/": Object {
            "addrs": "192.168.0.1",
            "rtts": "104457	78781",
            "service.address": "192.168.0.1",
            "service.id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
            "service.port": "0",
            "service.transport": "icmp",
            "ts": "1583271150",
          },
          "192.168.0.1/53/udp/": Object {
            "addrs": "192.168.0.1",
            "fp.certainty": "0.85",
            "hostname.bind": "router",
            "os.cpe23": "cpe:/o:canonical:ubuntu_linux:-",
            "os.family": "Linux",
            "os.product": "Linux",
            "os.vendor": "Ubuntu",
            "product": "ISC:BIND:9.11.3",
            "protocol": "dns",
            "replies": "www.google.com=A,172.217.1.132",
            "resolvers": "172.253.212.13",
            "rtts": "158454077	93812355",
            "service.address": "192.168.0.1",
            "service.cpe23": "cpe:/a:isc:bind:9.11.3",
            "service.family": "BIND",
            "service.id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
            "service.port": "53",
            "service.transport": "udp",
            "subnets": "1.2.3.0/24-f=1-c=0-s=0",
            "ts": "1583271194",
            "version.bind": "9.11.3-1ubuntu1.11-Ubuntu",
          },
          "192.168.0.1/5353/udp/": Object {
            "protocol": "mdns",
            "replies": "1.0.168.192.in-addr.arpa.=PTR,router.local.",
            "service.address": "192.168.0.1",
            "service.id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
            "service.port": "5353",
            "service.transport": "udp",
            "ts": "1583271140",
          },
        },
        "site_id": "c9f0bc41-afdb-4053-945a-2a5333d1565a",
        "site_name": "Corporate",
        "software_count": 0,
        "source_ids": Array [
          1,
        ],
        "sources": Array [
          "rumble",
        ],
        "subnets": Object {},
        "type": "Server",
        "updated_at": 1642795487,
        "vulnerability_count": 0,
      },
    },
  ],
  "_type": "rumble_asset",
  "alive": true,
  "category": "Server",
  "createdAt": 1642795487,
  "createdOn": undefined,
  "detectedBy": "arp",
  "deviceId": null,
  "displayName": "ROUTER",
  "firstSeen": 1583271138,
  "hardware": "",
  "hardwareModel": "",
  "hardwareVendor": "",
  "hardwareVersion": "",
  "id": "40fa9e9a-e798-44e4-a521-d06a7782ed14",
  "lastAgentId": "00000000-0000-0000-0000-000000000000",
  "lastSeen": 1583271232,
  "lastSeenOn": 1583271232,
  "lastTaskId": "374cc0f8-ca77-40d8-96c8-b53d563e7880",
  "macAddress": Array [
    "a0:36:9f:50:77:d8",
  ],
  "macAddresses": Array [
    "a0:36:9f:50:77:d8",
  ],
  "make": "",
  "model": "",
  "name": "ROUTER",
  "newestMacAddress": "a0:36:9f:50:77:d8",
  "newestMacVendor": "Intel Corporate",
  "orgName": "Demo Organization",
  "osDetails": "Ubuntu Linux",
  "osName": "Linux",
  "osVendor": "Ubuntu",
  "osVersion": "",
  "scanned": true,
  "serial": null,
  "serviceCount": 4,
  "siteName": "Corporate",
  "softwareCount": 0,
  "type": "Server",
  "updatedAt": 1642795487,
  "version": "",
  "vulnerabilityCount": 0,
}
`;
